src/main/java/com/commercetools/sync/producttypes/utils/ProductTypeUpdateAttributeDefinitionActionUtils.java

```diff
@@ -0,0 +1,279 @@
+     *                                                      update actions.
+     * @return a list of attribute definitions update actions if the list of attribute definitions is not identical.
+     *         Otherwise, if the attribute definitions are identical, an empty list is returned.
+     * @throws BuildUpdateActionException in case there are attribute definitions drafts with duplicate names.
+     */
+    @Nonnull
+    private static List<UpdateAction<ProductType>>
+        buildAttributeDefinitionUpdateActionsWithNewAttributeDefinitionDrafts(
+            @Nonnull final List<AttributeDefinition> oldAttributeDefinitions,
+            @Nonnull final List<AttributeDefinitionDraft> newAttributeDefinitionsDrafts,
+            @Nonnull final ProductTypeAttributeActionFactory productTypeAttributeDefinitionActionFactory)
+            throws BuildUpdateActionException {
+
+        final HashSet<String> removedAttributesDefinitionsNames = new HashSet<>();
+
+        final Map<String, AttributeDefinition> oldAttributesDefinitionsNameMap =
+            oldAttributeDefinitions
+                .stream()
+                .collect(toMap(AttributeDefinition::getName, attributeDefinition -> attributeDefinition));
+
+        final Map<String, AttributeDefinitionDraft> newAttributesDefinitionsDraftsNameMap;
+        try {
+            newAttributesDefinitionsDraftsNameMap = newAttributeDefinitionsDrafts.stream().collect(
+                toMap(AttributeDefinitionDraft::getName, attributeDefinitionDraft -> attributeDefinitionDraft,
+                    (attributeDefinitionDraftA, attributeDefinitionDraftB) -> {
+                        throw new DuplicateNameException("Attribute definitions drafts have duplicated names. "
+                            + "Attribute definitions names are expected to be unique inside their product type.");
+                    }
+                ));
+        } catch (final DuplicateNameException exception) {
+            throw new BuildUpdateActionException(exception);
+        }
+
+        final List<UpdateAction<ProductType>> updateActions =
+            buildRemoveAttributeDefinitionOrAttributeDefinitionUpdateActions(
+                oldAttributeDefinitions,
+                removedAttributesDefinitionsNames,
+                newAttributesDefinitionsDraftsNameMap,
+                productTypeAttributeDefinitionActionFactory
+            );
+
+        updateActions.addAll(
+            buildAddAttributeDefinitionUpdateActions(
+                newAttributeDefinitionsDrafts,
+                oldAttributesDefinitionsNameMap,
+                productTypeAttributeDefinitionActionFactory
+            )
+        );
+
+        buildChangeAttributeDefinitionOrderUpdateAction(
+            oldAttributeDefinitions,
+            newAttributeDefinitionsDrafts,
+            productTypeAttributeDefinitionActionFactory
+        )
+                .ifPresent(updateActions::add);
+
+        return updateActions;
```

Comment:
Since there is a lot of workflow in the method, How about adding inline comments in the method to describe each step? like the case with the `com.commercetools.sync.commons.utils.AssetsUpdateActionUtils#buildAssetsUpdateActionsWithNewAssetDrafts`

I think it would be useful for maintainers of the method.

I actually opened them next to each other to understand the algorithm :)

Reply:
I'm not a big fan of writing comments to explain "what". I think inline comments should explain "why" (i.e in the case you mentioned, you explained really well why `changeAssetOrder` actions are before `addAsset` actions, and it really helped me to understand the reason  👏), and code should speaks by itself through variables and method names 🙈. 

I always get confused with big inline comments since I can't read the code right away (in some cases there are more comments than code)

But that's a personal thing (I have seen a lot discussions around this topic)

Anyways, I add inline comments to explain the algorithm 😊

Reply:
@jortizsao I definitely agree with commenting "why" not the "what" :) The code should explain the "what" itself. I even decided to remove some unneeded inline comments in my price sync.